src/main/java/com/shapesecurity/shift/es2017/semantics/visitor/EqualityChecker.java

```diff
@@ -0,0 +1,37 @@
+package com.shapesecurity.shift.es2017.semantics.visitor;
+
+import com.shapesecurity.functional.data.ImmutableList;
+import com.shapesecurity.shift.es2017.semantics.asg.Node;
+
+import javax.annotation.Nonnull;
+import java.util.Iterator;
+
+public final class EqualityChecker {
+
+	private EqualityChecker() {
+
+	}
+
+	public static boolean nodesAreEqual(@Nonnull Node node1, @Nonnull Node node2) {
+		if (node1 == node2) {
+			return true;
+		}
+		if (!node1.partiallyEquals(node2)) {
+			return false;
+		}
+		ImmutableList<Node> nodeList1 = Visitor.getChildren(node1);
+		ImmutableList<Node> nodeList2 = Visitor.getChildren(node2);
+		if (nodeList1.length != nodeList2.length) {
+			return false;
+		}
+		Iterator<Node> node1Iterator = nodeList1.iterator();
+		Iterator<Node> node2Iterator = nodeList2.iterator();
+		while (node1Iterator.hasNext() && node2Iterator.hasNext()) {
+			if (!node1Iterator.next().partiallyEquals(node2Iterator.next())) {
```

Comment:
This isn't sufficient, surely? `x = 1 + 2 + 3` should not be equal to `x = 1 + 2 + 4`.